### ISSUE TITLE
Adding operations after gradient are applied screws with the optimizer

### ISSUE THREAD

**Issue by domluna** [tensorflow/tensorflow#412]
I'm currently trying to add an operation after I apply the gradients. Essentially I have a row of some variables that I don't to change via gradients so I would ideally reset it after gradients are applied (I could also set that part of the gradient to 0 but w/e). I currently have this done after my forward inference pass as putting it after apply_gradients doesn't seem to work.

Even adding an op like tf.Print seems to completely screw with the optimizer. What I gather from my debugging is no variables are updated.

```
train_op = self._opt.apply_gradients(...)
train_op = tf.Print(train_op, ....)
```
If I understand Print correctly it just prints stuff and then returns have the op that was passed in, so the returned op should work the same. tf.Print works as expected it prints out what I wanted it to print.

**Comment by girving** [comment]
How are you using `train_op` downstream?  What you're doing seems like it should work as you expect from what you've written, so I think we need to see the rest of the context.

**Comment by vrv** [comment]
Closing due to lack of activity / follow up.  Feel free to comment and we'll re-open.

**Comment by asaydin** [comment]
I think I am having a similar problem to this. Gradients are apparently computed, but the weight matrix associated with the gradient does not change (i.e., they are exactly the same across mini-batches and iterations). Interestingly, this happens only with one dataset. I use the same code for two other (MNIST and an artificially generated dataset) datasets, and it works fine. 

```
optimizer = tf.train.GradientDescentOptimizer(0.1)
predict_op = tf.argmax(py_x,1)

tvars = tf.trainable_variables()
grads = tf.gradients(cost, tvars)
train_op = optimizer.apply_gradients(zip(grads, tvars))
```

So this is my optimization code. I used this explicit way of computing to see the values of the gradients. Otherwise I used

`train_op = tf.train.GradientDescentOptimizer(0.1).minimize(cost)`

and other optimization techniques TensorFlow offers as well. As I said, although gradients are computed(And they are non-zero), they are not applied to the weight matrix, and the weight matrix remains the same. As a result, because of the random initialization of the weights, the classifier I trained (Which is an LSTM + a softmax layer) keeps predicting with random guess (i.e., 1/n accuracy, where n is the number of classes). 

Not sure if this is the right place to post, but I guess the problem is similar. I can open a new issue if you want. The code is very long (and messy because of everything I tried to make it work in the last few days), so I did not post here the whole thing. If you need the other parts, I can clean them and post here.

**Comment by girving** [comment]
Are you running the `train_op`?

**Comment by asaydin** [comment]
My bad. I deleted it at some point, and only for the code that runs with that particular dataset. Thanks a lot. You saved my future few days.
